components/layout/traversal.rs

```diff
@@ -185,8 +185,8 @@ impl<'a> PreorderDomTraversal for RecalcStyleForNode<'a> {
 
                     if node.as_element().is_some() {
                         // Perform the CSS selector matching.
-                        let stylist = unsafe { &*self.layout_context.shared.stylist };
-                        node.match_node(stylist,
+                        let stylist = self.layout_context.shared.stylist.read().unwrap();
```

Comment:
It would be nice if we could do that. I'm fine with changing the WorkQueue API.

I think taking a lock for every node here will probably erase most parallelism in Servo, so we can't land this.